compiler/lib/effects.ml

```diff
@@ -482,63 +557,86 @@ let cps_last ~st ~alloc_jump_closures pc (last : last) ~k : instr list * last =
+      let cps_params, cps_cont = Hashtbl.find st.closure_info pc in
       st.in_cps := Var.Set.add x !(st.in_cps);
-      Let (x, Closure (params @ [ k ], cont))
+      [ Let (direct_c, Closure (params, cont))
+      ; Let (cps_c, Closure (cps_params, cps_cont))
+      ; Let (x, Prim (Extern "caml_cps_closure", [ Pv direct_c; Pv cps_c ]))
+      ]
   | Let (x, Prim (Extern "caml_alloc_dummy_function", [ size; arity ])) -> (
       match arity with
       | Pc (Int a) ->
-          Let
-            ( x
-            , Prim
-                (Extern "caml_alloc_dummy_function", [ size; Pc (Int (Targetint.succ a)) ])
-            )
+          [ Let
+              ( x
+              , Prim
+                  ( Extern "caml_alloc_dummy_function"
+                  , [ size; Pc (Int (Targetint.succ a)) ] ) )
+          ]
       | _ -> assert false)
   | Let (x, Apply { f; args; _ }) when not (Var.Set.mem x st.cps_needed) ->
       (* At the moment, we turn into CPS any function not called with
          the right number of parameter *)
-      assert (Global_flow.exact_call st.flow_info f (List.length args));
-      Let (x, Apply { f; args; exact = true })
+      assert (
+        (* If this function is unknown to the global flow analysis, then it was
+           introduced by the lambda lifting and does not require CPS *)
+        Var.idx f >= Var.Tbl.length st.flow_info.info_approximation
+        || Global_flow.exact_call st.flow_info f (List.length args));
+      [ Let (x, Apply { f; args; exact = true }) ]
   | Let (_, (Apply _ | Prim (Extern ("%resume" | "%perform" | "%reperform"), _))) ->
+      (* Applications of CPS functions and effect primitives require more work
+         (allocating a continuation and/or modifying end-of-block branches) and
+         are handled in a specialized function below. *)
       assert false
-  | _ -> instr
+  | Let (x, Prim (Extern "caml_assume_no_perform", [ Pv f ])) ->
+      if double_translate ()
+      then
+        (* We just need to call [f] in direct style. *)
+        let unit = Var.fresh_n "unit" in
+        let exact = Global_flow.exact_call st.flow_info f 1 in
+        [ Let (unit, Constant (Int Targetint.zero))
+        ; Let (x, Apply { exact; f; args = [ unit ] })
+        ]
+      else (
+        (* The "needs CPS" case should have been taken care of by another, specialized
+           function below. *)
+        assert (not (Var.Set.mem x st.cps_needed));
+        (* Translated like the [Apply] case, with a unit argument *)
+        assert (
+          (* If this function is unknown to the global flow analysis, then it was
+             introduced by the lambda lifting and does not require CPS *)
+          Var.idx f >= Var.Tbl.length st.flow_info.info_approximation
+          || Global_flow.exact_call st.flow_info f 1);
+        let unit = Var.fresh_n "unit" in
+        [ Let (unit, Constant (Int Targetint.zero))
+        ; Let (x, Apply { f; args = [ unit ]; exact = true })
+        ])
```

Comment:
This case cannot happen since the partial CPS analysis enforces a CPS transform for this primitive.